---
title: "Rqtl_SARS-CoV-1"
author: "Ellen Risemberg"
date: "9/14/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Environment setup 

```{r message=FALSE, warning=FALSE}
library(knitr)
library(qtl)
library(qtl2)
library(MESS)
library(MASS)
library(ggplot2)
library(extRemes)
source("code-dependencies/qtl_functions.R")
source("code-dependencies/plot_noX.R")
```

Load ggplot themes:

```{r}
load_themes()
```

## Load data 

```{r}
SARS <- read.cross(format="csv", file='derived_data/rqtl_files/SARS_CC006xCC044_rqtl.csv', 
                   na.strings=c("-","NA", "not tested"), genotypes=c("AA","AB","BB"), 
                   alleles=c("A","B")) # A = CC006, B = CC044
SARS <- jittermap(SARS)
summary(SARS)
plotMap(SARS) 

SARS$pheno$batch <- as.factor(SARS$pheno$batch) # change batch to factor
```

Genetic map statistics:

```{r}
SARSv2 <- convert2cross2(SARS) # convert to Rqtl2 cross 
find_map_gaps(SARSv2$gmap, min_gap=10)
```

Median physical distance in Mbp:

```{r distance}
alldistances <- vector()
  
for (i in 1:length(SARSv2$gmap)){
  positions <- SARSv2$gmap[[i]]
  distances <- positions[2:length(positions)] - positions[1:(length(positions)-1)]
  
  alldistances <- append(alldistances, distances)
}

median(alldistances)
```

## Calculate derived measures

**Calculate percentage of starting weight**

```{r}
SARS$pheno$pd0 <- 100
SARS$pheno$pd1 <- (SARS$pheno$d1/SARS$pheno$d0)*100
SARS$pheno$pd2 <- (SARS$pheno$d2/SARS$pheno$d0)*100
SARS$pheno$pd3 <- (SARS$pheno$d3/SARS$pheno$d0)*100
SARS$pheno$pd4 <- (SARS$pheno$d4/SARS$pheno$d0)*100
```

**Calculate area under the curve**

Use `MESS::auc` function which calculates the AUC for the weight trajectory, then subtract that from 400 (total area beneath 100% line from 0-4). Any weight above 100% would increase the AUC and decrease the area-above-the-curve (so some AACs will be negative).

Example: 

```{r}
plot(c(0,1,2,3,4), SARS$pheno[1,c('pd0', 'pd1', 'pd2', 'pd3', 'pd4')], 
     xlab = "Day post-infection", ylab = "% of starting weight")
lines(c(0,1,2,3,4), SARS$pheno[1,c('pd0', 'pd1', 'pd2', 'pd3', 'pd4')])
abline(h=100)
```

Metric calculation:

```{r}
SARS <- calc_auc(cross = SARS, steps = c(0,1,2,3,4), col.name = "weight_aac", 
                 phenos = c('pd0', 'pd1', 'pd2', 'pd3', 'pd4'))
```


## Exploratory data analysis 

How much weight do SARS-infected mice lose on average? 

```{r}
pctloss <- (1-(SARS$pheno$d4/SARS$pheno$d0))*100
mean(pctloss)
range(pctloss)
```

Plot weight trajectory with CC044 and CC006 avg trajectory: 

```{r}
wtloss <- cov_trajectory_plot(SARS, phenos = c('pd0', 'pd1', 'pd2', 'pd3', 'pd4'), 
                              title = "SARS-CoV MA15", ylab = "% of starting weight", 
                              ylim = c(72,114))
wtloss
```

Save plot:

```{r}
ensure_directory("figures/SARS1")
png(filename = "figures/SARS1/wt_loss.png", width = 700)
wtloss + bw_big_theme
dev.off()
```

Save for combining with other figures:

```{r}
wtloss <- wtloss + theme(legend.position = "none") + bw_big_theme
ensure_directory("derived_data/otherRobjects")
saveRDS(wtloss, file = "derived_data/otherRobjects/sars1wtloss.Rdata")
```

### Hemorrhage score

Average HS:

```{r}
mean(SARS$pheno$HS)
range(SARS$pheno$HS)
median(SARS$pheno$HS)
```


```{r}
plotPheno(SARS, pheno.col = 'HS')
```


## Load model info table

```{r}
models <- read.csv("source_data/SARS1-models.csv", na.strings = "")
```

## Transform data 

Before transforming data, save raw phenotypes for plotting later:

```{r}
raw_phenos <- SARS$pheno
```

Log-transform weight on d1-d4, don't transform d0 (covariate only), AAC (already normal), or HS (categorical). 

```{r}
SARS$pheno$d1 <- log(SARS$pheno$d1)
SARS$pheno$d2 <- log(SARS$pheno$d2)
SARS$pheno$d3 <- log(SARS$pheno$d3)
SARS$pheno$d4 <- log(SARS$pheno$d4)
``` 

## Covariate analysis 

anova test - which of the covariates/interactions have a significant effect?

```{r}
anova(aov(d1 ~ d0 + batch + sex, data = SARS$pheno))
anova(aov(d2 ~ d0 + batch + sex, data = SARS$pheno))
anova(aov(d3 ~ d0 + batch + sex, data = SARS$pheno))
anova(aov(d4 ~ d0 + batch + sex, data = SARS$pheno))
anova(aov(weight_aac ~ batch + sex, data = SARS$pheno))
anova(aov(HS ~ batch + sex, data = SARS$pheno))
```

Sex, batch and baseline weight all affect our phenotypes (batch and sex only for AAC and HS).

## Define covariates 

Covariates: 
* Sex 
* Batch (as string)
* Cross direction? Use parent_info.xlsx 

There are eight batches - probably don't need to model as a random effect? 

```{r}
covar <- cbind(as.numeric(pull.pheno(SARS, "sex") == "M"),
               as.numeric(pull.pheno(SARS, "batch") == 2), # batch 1 is baseline
               as.numeric(pull.pheno(SARS, "batch") == 3),
               as.numeric(pull.pheno(SARS, "batch") == 4),
               as.numeric(pull.pheno(SARS, "batch") == 5),
               as.numeric(pull.pheno(SARS, "batch") == 6),
               as.numeric(pull.pheno(SARS, "batch") == 7),
               as.numeric(pull.pheno(SARS, "batch") == 11),
               pull.pheno(SARS, 'd0'))
colnames(covar) <- c('sex', 'batch2', 'batch3', 'batch4', 'batch5', 'batch6', 'batch7', 'batch11', 'd0')
```


# Single-QTL Analysis 

First, run `calc.genoprob()`:

```{r}
SARS <- calc.genoprob(SARS)
```

`models` dataframe has the following info about each phenotype:
* `obj`: name of `scanone` object 
* `perm.obj`: name of `scanoneperm` object 
* `type`: model type (normal, np, 2-part)
* `colname`: column name in input spreadsheet 
* `name`: name for genome scan title
* `abbr`: name for axis labels 

Model types:
* Lung HS: nonparametric model (categorical variable)
* Weight loss (% weight lost): normal model 
* Weight loss (derived measures): normal model 

```{r}
kable(models)
```

Create models (covariates will be ignored for non-parametric and two-part model):

```{r}
ensure_directory("derived_data/SARS1")
ensure_directory("derived_data/SARS1/mods")
create_models(cross.obj = SARS, models = models, covar = covar, mod.dir = "derived_data/SARS1/mods/")
```

Create permutations: 

```{r}
ensure_directory("derived_data/SARS1/perms")
create_perms(cross.obj = SARS, models = models, perm.dir = "derived_data/SARS1/perms/", 
             perm.Xsp = TRUE)
```

Plot genome scans: 

```{r}
plot_scans(models = models)
```

Save genome scans:

```{r}
ensure_directory("figures/SARS1/scans")
plot_scans(models, save = TRUE, save.dir = 'figures/SARS1/scans/')

ensure_directory('figures/SARS1/scans/same_ylim/')
plot_scans(models, save = TRUE, ylim = c(0,8.3), save.dir = 'figures/SARS1/scans/same_ylim/')
```


## Significant QTLs 

Calculate 95% Bayes credible intervals for significant LOD peaks, and create a summary table with data for each LOD peak: marker, chromosome, position, LOD and positions of Bayes credible intervals.

```{r}
peaks <- doc_peaks(models, sig.level = 0.10)
kable(peaks)
```

### Table for manuscript 

Turn into tibble for easier processing:

```{r}
peakstbl <- as_tibble(peaks) %>% arrange(factor(chr, levels = c('9', '19', '3', '2')))
```


### Adjusted P-values

```{r message=FALSE}
peakstbl$adj_pval <- rep(NA, nrow(peakstbl))

for (i in 1:nrow(peakstbl)){
  modname <- peakstbl$model[i]
  permname <- models$perm.obj[which(models$obj == modname)]
  chr <- peakstbl$chr[i]
  
  # Don't use this method as it can result in p = 0 if no permuted LODs 
  # are greater than the observed LOD 
  # p <- mean(get(permname)$A > summary(get(modname))[as.integer(chr),'lod'])
  
  fitgev <- fevd(as.numeric(get(permname)$A), type = "GEV")
  fitgevsum <- summary.fevd(fitgev)
  
  p <- pevd(q = as.numeric(peakstbl$lod[i]), 
            loc = fitgevsum$par[1], 
            scale = fitgevsum$par[2], 
            shape = fitgevsum$par[3], 
            lower.tail = FALSE)

  
  peakstbl$adj_pval[i] <- p
}

peakstbl$adj_pval <- format(peakstbl$adj_pval, digits = 2, scientific = TRUE)
```


### Genomic location 

Create chr:position(interval) column. First, split up bayes CI column into lower and upper limit. Then concatenate various data to produce column:

```{r}
peakstbl$BayesCIlower <- rep(NA, nrow(peakstbl))
peakstbl$BayesCIupper <- rep(NA, nrow(peakstbl))

for (i in 1:nrow(peakstbl)){
  peakstbl$BayesCIlower[i] <- strsplit(peakstbl$`Bayes CI`[i], split = ' - ')[[1]][1]
  peakstbl$BayesCIupper[i] <- strsplit(peakstbl$`Bayes CI`[i], split = ' - ')[[1]][2]
}

# Then concatenate  
peakstbl$chrint <- paste0(rep('Chr ',nrow(peakstbl)), 
                 peakstbl$chr, 
                 rep(': ',nrow(peakstbl)),
                 trimws(format(round(as.double(peakstbl$pos), 2), nsmall=2)),
                 rep(' (',nrow(peakstbl)),
                 trimws(format(round(as.double(peakstbl$BayesCIlower), 2), nsmall=2)),
                 rep('-',nrow(peakstbl)),
                 trimws(format(round(as.double(peakstbl$BayesCIupper), 2), nsmall=2)),
                 rep(')',nrow(peakstbl)))
```

Use phenotype name instead of model name: 

```{r}
peakstbl$Phenotype <- models$name[match(peakstbl$model, models$obj)]
```


### Unadjusted P-values 

```{r}
peakstbl$unadj_pval <- rep(NA, nrow(peakstbl))

for (i in 1:nrow(peakstbl)){
  marker = peakstbl$marker[i]
  mod = peakstbl$model[i]
  pheno.col = models$colname[models$obj == mod]
  covar_names <- models$cov[models$obj == mod]
  if (covar_names == "") {covar_names <- NULL}
  
  unadj.pval <- get_unadj_pval(SARS, pheno.col, marker, covar_df = covar, covar_names = covar_names)
  peakstbl$unadj_pval[i] <- unadj.pval
}

peakstbl$unadj_pval <- format(peakstbl$unadj_pval, digits = 3)
```


### QTL names 

```{r}
peakstbl$QTL <- rep(NA, nrow(peakstbl))

qtl_pheno_counts <- peakstbl %>% group_by(chr) %>% dplyr::summarize(n = n())

peakstbl$QTL <- c(rep('HrS43', qtl_pheno_counts$n[qtl_pheno_counts$chr==9]),
                  rep('HrS46', qtl_pheno_counts$n[qtl_pheno_counts$chr==19]), 
                  rep('HrS48', qtl_pheno_counts$n[qtl_pheno_counts$chr==3]),
                  rep('HrS49', qtl_pheno_counts$n[qtl_pheno_counts$chr==2]))
```


### Phenotypic variance explained


AKA heritability due to a QTL. Using formula (described in Rqtl guide pg 122)

$$h^2 = \frac{\text{var}\{E(y|g\}}{\text{var}\{y\}}$$

For F2 cross: 

$$h^2 = \frac{2a^2 + d^2}{2a^2 + d^2+4\sigma^2}$$

where $a = \mu_{BB} - \mu_{AA}/2$, $d=\mu_{AB} - (\mu_{AA}+\mu_{BB})/2$, and $\sigma^2$ is the residual variance. 

```{r}
peakstbl$phenotypic.var.expl <- rep(NA, nrow(peakstbl))

for (i in 1:nrow(peakstbl)){
  marker = peakstbl$marker[i]
  mod = peakstbl$model[i]
  pheno.col = models$colname[models$obj == mod]
  covar_names <- models$cov[models$obj == mod]
  if (covar_names == "") {covar_names <- NULL}
  
  h2 <- get_var_expl(SARS, pheno.col, marker, covar_df = covar, covar_names = covar_names)
  
  peakstbl$phenotypic.var.expl[i] <- h2
}

peakstbl$phenotypic.var.expl <- paste0(as.numeric(format(peakstbl$phenotypic.var.expl, digits=3))*100, '%')
```


### Analysis done, format and print

Remove redundant data, re-order: 

```{r}
table1 <- peakstbl %>% 
  arrange(match(chr, c('9', '19', '3', '2')), as.numeric(adj_pval))

# now that it's sorted by adjusted p-value, add */**/*** (makes it a string so can't sort)
table1$adj_pval <- paste(table1$adj_pval, ifelse(as.numeric(table1$adj_pval) < 0.005, '(***)',
                                                 ifelse(as.numeric(table1$adj_pval) < 0.05, '(**)',
                                                        ifelse(as.numeric(table1$adj_pval) < 0.10, '(*)', NA))))

col_order <- c('QTL', 'Phenotype', 'chrint', 'adj_pval', 'phenotypic.var.expl')
table1 <- table1[,col_order]
```

Save to csv: 

```{r}
ensure_directory("results")
write.csv(table1, file = "results/SARS1-QTLsummary.csv", row.names = FALSE)
```


## Phenotype x genotype plots for markers at significant LOD peaks 

Name QTLs for PxG plotting

```{r}
qtl.map <- data.frame(chr = c(9, 19, 3, 2), 
                      qtl_name = c('HrS43', 'HrS46', 'HrS48', 'HrS49'))
```

Print all pxg plots: 

```{r}
plot_pxg(cross.obj = SARS, cross.type = 'f2', raw.data = raw_phenos,
         geno.map = list(A = "CC006", B = "CC044"), qtl.map = qtl.map,
         peaks = peaks, plot.type = 'pxg', theme = rmd_theme)
```

Save PxG plots:

```{r}
ensure_directory("figures/SARS1/pxg")
plot_pxg(SARS, cross.type = 'f2', raw.data = raw_phenos, 
         geno.map = list(A = "CC006", B = "CC044"), qtl.map = qtl.map, 
         peaks = peaks, plot.type = 'pxg', theme = rmd_theme, save = TRUE, 
         save.dir = 'figures/SARS1/pxg/')
```

PxG plot for Figure 4:

```{r}
HrS43m <- 'S3N094839317'
p <- pxg(cross = SARS, pheno = SARS$pheno$weight_aac, marker = HrS43m,
         qtl.map = qtl.map, geno.map = list(A = "CC006", B = "CC044"), 
         theme = sbs_pxg_theme, title = "SARS-CoV MA15", ylab = "Weight loss (AAC)", 
         ylim = c(-20,60))
```

Save plot:

```{r}
ensure_directory("figures/Fig4")
png("figures/Fig4/sars1-HrS43-aac.png", width = 600)
p
dev.off()
```

Save R object for combining with other plots:

```{r}
saveRDS(p, file = "derived_data/otherRobjects/sars1-pxg-HrS43-aac.Rdata")
```


## Multiple QTL Analysis

Are chr2, chr3, chr19 significant after controlling for chr9?

```{r}
geno2 <- pull.geno(SARS, 2)[,'gUNC2810208']
geno3 <- pull.geno(SARS, 3)[,'gUNC6002742']
geno9 <- pull.geno(SARS, 9)[,'S3N094839317']
geno19 <- pull.geno(SARS, 19)[,'gUNC30234572']

# significant on their own:
lmod <- lm(d4 ~ d0 + geno2, data = SARS$pheno)
summary(lmod)

lmod2 <- lm(d4 ~ d0 + geno3, data = SARS$pheno)
summary(lmod2)

lmod3 <- lm(d4 ~ d0 + geno9, data = SARS$pheno)
summary(lmod3)

lmod4 <- lm(d4 ~ d0 + geno19, data = SARS$pheno)
summary(lmod4)

# significant together: 
lmodfull <- lm(d4 ~ d0 + geno9 + geno2 + geno3 + geno19, data = SARS$pheno)
summary(lmodfull)
```


